src/waveform/widgets/allshader/waveformwidget.cpp

```diff
@@ -117,31 +135,30 @@ mixxx::Duration WaveformWidget::render() {
 }
 
 void WaveformWidget::paintGL() {
-    if (shouldOnlyDrawBackground()) {
-        if (!m_rendererStack.empty()) {
-            m_rendererStack[0]->allshaderWaveformRenderer()->paintGL();
-        }
-    } else {
-        for (auto* pRenderer : std::as_const(m_rendererStack)) {
-            pRenderer->allshaderWaveformRenderer()->paintGL();
-        }
-    }
+    // opacity of 0.f effectively skips the subtree rendering
+    m_pOpacityNode->setOpacity(shouldOnlyDrawBackground() ? 0.f : 1.f);
+
+    m_pEngine->preprocess();
+    m_pEngine->render();
 }
 
 void WaveformWidget::castToQWidget() {
     m_widget = this;
 }
 
 void WaveformWidget::initializeGL() {
-    for (auto* pRenderer : std::as_const(m_rendererStack)) {
-        pRenderer->allshaderWaveformRenderer()->initializeGL();
-    }
+}
+
+void WaveformWidget::resizeRenderer(int, int, float) {
+    // defer to resizeGL
 }
```

Comment:
what do you mean `defer`? Does this assume `resizeGL` is called anyways shortly after?

Reply:
I don't remember. I will investigate and improve the comment.

Reply:
yeah, this is called when the widget is resized, but in the case of wglwidgets we also get the resizeGL call, which is more useful as it has the right context. I'll clarify in the comment.